Gear/joh 54 johnjud gateway image endopint

macgeargear opened the pull request:
## Change made

- [ ]  Bug fixes
- [x]  New features
- [ ]  Breaking changes
## Describe what you have done
- add image service & handler (no test)

### New Features
- image end point

### Fix

### Others

macgeargear committed:
chore

macgeargear committed:
chore

macgeargear committed:
fix: image dto

macgeargear committed:
fix: image response dto

macgeargear committed:
chore

macgeargear committed:
chore

macgeargear committed:
feat: image handler

macgeargear committed:
feat: image service

macgeargear committed:
feat: image utils

macgeargear committed:
gen mock

macgeargear committed:
fix: image service interface

macgeargear committed:
chore

macgeargear committed:
chore

macgeargear committed:
chore

macgeargear committed:
chore

macgeargear committed:
chore

bookpanda commented on src/app/service/image/image.service.go:
```diff
@@ -15,14 +24,115 @@ func NewService(client proto.ImageServiceClient) *Service {
+func (s *Service) Delete(id string) (*dto.DeleteImageResponse, *dto.ResponseErr) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	request := &proto.DeleteImageRequest{
```
I think we should add a blank line here

bookpanda commented on src/app/handler/image/image.handler.go:
```diff
@@ -1,28 +1,100 @@
 func (h *Handler) Upload(c *router.FiberCtx) {
+	request := &dto.UploadImageRequest{}
+	err := c.Bind(request)
+	fmt.Println("request: ", request)
```
Did you forget to remove?

bookpanda reviewed (changes requested):
Actually, Image needs one more endpoint: AssignPet: POST /image/assign/:pet_id		
(FindByPetId doesn't have to be an endpoint as we'll call it in backend's pet service, but you don't have to remove the code).

Also, I think ```main.go``` still doesn't have these set as endpoints.
Please also add Upload, Delete, AssignPet paths to ```AdminPath``` and FindByPetId to ```ExcludePath``` in ```auth.constant.go```

macgeargear committed:
chore

macgeargear committed:
feat: get image path

macgeargear committed:
chore

macgeargear committed:
feat: image handler

macgeargear committed:
chore

macgeargear commented:
> Actually, Image needs one more endpoint: AssignPet: POST /image/assign/:pet_id (FindByPetId doesn't have to be an endpoint as we'll call it in backend's pet service, but you don't have to remove the code).
> 
> Also, I think `main.go` still doesn't have these set as endpoints. Please also add Upload, Delete, AssignPet paths to `AdminPath` and FindByPetId to `ExcludePath` in `auth.constant.go`

the end point you want me to remove in `main.go` is just `FindByPetId` and the rest still the same right?, and one more thing, how can i signup as admin because when i want to test api call e.g. `POST pet` it'll block by `auth middleware` (i have to comment keys `AdminPath` in `auth.constant.go`)

macgeargear committed:
feat: dto

macgeargear committed:
feat: assign pet handler

macgeargear committed:
feat: assign pet service

macgeargear committed:
feat: admin path

macgeargear committed:
mock gen

macgeargear committed:
feat: image service interface

bookpanda commented on src/app/dto/image.dto.go:
```diff
@@ -1,7 +1,26 @@
+type AssignPetRequest struct {
+	Ids   []string `json:"ids"`
+	PetId string   `json:"pet_id"`
+}
```
We should add some validations here

bookpanda commented on src/app/handler/image/image.handler.go:
```diff
@@ -1,28 +1,133 @@
+		return
+	}
 
+	c.JSON(http.StatusCreated, response)
```
let's make it statusOK (it's update)

bookpanda commented on src/app/service/image/image.service.go:
```diff
@@ -15,14 +24,159 @@ func NewService(client proto.ImageServiceClient) *Service {
+	}, nil
+}
+
+func (s *Service) AssignPet(in *dto.AssignPetRequest) (*dto.AssignPetResponse, *dto.ResponseErr) {
```
AssignPet also have codes.InvalidArgument (when petId is not uuid)

bookpanda replied:
the code is in jj-file repo

bookpanda commented on src/constant/auth/auth.constant.go:
```diff
@@ -16,6 +16,10 @@ var AdminPath = map[string]struct{}{
 	"PUT /pets/:id/visible": {},
 	"DELETE /pets/:id":      {},
 	//need to add image upload, delete, assignpet
```
you can remove the comment

bookpanda commented on src/main.go:
```diff
@@ -142,6 +144,10 @@ func main() {
+	r.GetImage("/", imageHandler.FindByPetId)
+	r.PostImage("/", imageHandler.Upload)
+	r.DeleteImage("/:id", imageHandler.Delete)
+
```
we also have AssignPet

bookpanda commented:
> > Actually, Image needs one more endpoint: AssignPet: POST /image/assign/:pet_id (FindByPetId doesn't have to be an endpoint as we'll call it in backend's pet service, but you don't have to remove the code).
> > Also, I think `main.go` still doesn't have these set as endpoints. Please also add Upload, Delete, AssignPet paths to `AdminPath` and FindByPetId to `ExcludePath` in `auth.constant.go`
> 
> the end point you want me to remove in `main.go` is just `FindByPetId` and the rest still the same right?, and one more thing, how can i signup as admin because when i want to test api call e.g. `POST pet` it'll block by `auth middleware` (i have to comment keys `AdminPath` in `auth.constant.go`)

Yes, remove only FindByPetId and add AssignPet. 

You can signup, then signin, get the token, add it in auth section of postman (bearer token), no need to add ```Bearer ``` in front of token

macgeargear committed:
chore

macgeargear committed:
feat: assign pet

macgeargear committed:
fix: invalidArgument error

macgeargear committed:
feat: assign pet path

I think no need to use `/assign/:pet_id` because they're already in body request

macgeargear committed:
chore

macgeargear committed:
chore

bookpanda commented on src/app/dto/image.dto.go:
```diff
@@ -1,7 +1,26 @@
+
+type AssignPetRequest struct {
+	Ids   []string `json:"ids"`
+	PetId string   `json:"pet_id"`
```
Did you forget to push?

bookpanda commented on src/app/router/image.router.go:
```diff
@@ -15,3 +15,17 @@ func (r *FiberRouter) DeleteImage(path string, h func(ctx *FiberCtx)) {
+		h(NewFiberCtx(c))
+		return nil
+	})
+}
```
Actually, you can use ```PostImage```. It's post as in POST method

macgeargear committed:
chore

bookpanda commented on src/database/seeds/constant.seed.go:
```diff
@@ -0,0 +1,282 @@
+	{
+		Type:         "แมว",
+		Name:         "เต้าเจี้ยว",
+		Birthdate:    "4 July 2023",
```
I think we need to use the datetime format for the age filter to work.

"2024-01-19 17:14:55.513579" = time representation in the format of "YYYY-MM-DD HH:MM:SS.SSSSSS" (Year-Month-Day Hour:Minute:Second.Microsecond)

bookpanda replied:
you can leave the hour, minute, second, microsecond to 0 I suppose
